### PR TITLE
Local `tox` checks do not skip

### LINKED ISSUE #28550: Implement Pyproject.toml feedback
- [ ] Ignore `active/inactive` for every tox check when running locally. Local runs should always run as they are explicitly requested.
- [ ] CI should still honor inactive/active from pyproject.toml
- [ ] An analyze step that has a skip registered should be turned yellow in the analyze stage
- [ ] An analyze step which would have no packages run should be **entirely** skipped

### PR DESCRIPTION
**Pull request by scbedd**
Resolves #28550

**tl;dr** if all the checks in a static analyze are skipped, we should skip that step entirely.

This PR:

- [x] Updates `run_<check>` scripts to always run locally, regardless of whether or not a package is ignored. You can get the same behavior as CI by populating environment variable `TF_BUILD`.
- [x] Individual skips should turn step yellow
- [x] All packages skipping the package should result in that step not running at all. (applies to pylint, verifywhl, verifysdist, bandit, mypy)

Detailed description:

The job yaml itself supports some [global level skips](https://github.com/Azure/azure-sdk-for-python/blob/main/doc/eng_sys_checks.md#skipping-entire-sections-of-builds). We can take advantage of this feature by placing a script invocation at the beginning of the `Analyze` build. This script should evaluate the packages of a service directory, and if all skip a given check that is also associated with a global skip (EG: `Skip.Pylint`), that variable will be automatically set.

### CODE REVIEW

File: eng/tox/run_mypy.py [hunk 1]
```diff
@@ -61,24 +62,25 @@
...
+    if in_ci():
+        if not is_check_enabled(args.target_package, "type_check_samples", True):
             logging.info(
-                f"Running mypy commands on sample code: {sample_code}"
+                f"Package {package_name} opts-out of mypy check on samples."
             )
-            check_call(sample_code)
-        except CalledProcessError as sample_err:
-            sample_code_error = sample_err
+        else:
```
Comment (scbedd) [inline eng/tox/run_mypy.py:70]: Need to restructure this. It improperly makes it so that the sample code is only mypy-ed in CI, which is not the intent of this PR.

File: scripts/devops_tasks/set_tox_environment.py [hunk 2]
```diff
@@ -18,47 +23,77 @@
...
 
-def remove_unsupported_values(selected_set, unsupported_values):
+def remove_unsupported_values(selected_set: List[str], unsupported_values: List[str]):
     for unsupported_tox_env in unsupported_values:
         if unsupported_tox_env in selected_set:
             selected_set.remove(unsupported_tox_env)
 
 
+def process_ci_skips(service: str) -> None:
+    checks_with_global_skip = ["pylint", "verifywhl", "verifysdist" "bandit", "mypy"]
```
Comment (kristapratico) [inline scripts/devops_tasks/set_tox_environment.py:46]: Should this list include pyright, verifytypes?

File: tools/azure-sdk-tools/ci_tools/environment_exclusions.py [hunk 3]
```diff
@@ -345,6 +349,14 @@
 
 
 def is_check_enabled(package_path: str, check: str, default: bool = True) -> bool:
+    """
+    Single-use function to evaluate whether or not a given check should run against a package.
+
+    In order:
+     - Checks <CHECK>_OPT_OUT for package name.
+     - Honors override variable if one is present: <PACKAGE-NAME>_<CHECK>.
+     - Finally falls back to the pyproject.toml at package root (if one exists) for a tools setting enabling/disabling <check>.
```
Comment (kristapratico) [inline tools/azure-sdk-tools/ci_tools/environment_exclusions.py:358]: sorry if dumb - what's the reason for keeping the CHECK_OPT_OUT lists if we have the pyproject.toml? Do they capture different information?

Comment (scbedd) [reply tools/azure-sdk-tools/ci_tools/environment_exclusions.py:358]: Because I haven't scripted out their removal in favor of individual pyproject.toml.

File: scripts/devops_tasks/set_tox_environment.py (hunk 2 above)
Comment (scbedd) [reply scripts/devops_tasks/set_tox_environment.py:46]: The way I was thinking about this originally, I would only apply this logic if the global skip variable is honored. EG we actually have steps that honor `Skip.Pyright`. We _do_ have a step that conditions on `$(Skip.Pylint)`, but we don't on those two you don't mention.

I am going to update this such that it always sets the skip variable, even if it doesn't do anything necessarily.

**Review by kristapratico — APPROVED** [review]

File: tools/azure-sdk-tools/ci_tools/ci_interactions.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import os, inspect
+from ci_tools.variables import in_ci
+
+
+def output_ci_warning(message: str, location=None) -> None:
+    ci_type = in_ci()
```
Comment (mccoyp) [inline tools/azure-sdk-tools/ci_tools/ci_interactions.py:6]: I understand why this is being done, but it does seem a little weird. I would consider making the `TF_BUILD`/`CI` check come from a different method

**Review by mccoyp — APPROVED** [review]

File: tools/azure-sdk-tools/ci_tools/ci_interactions.py (hunk 4 above)
Comment (scbedd) [reply tools/azure-sdk-tools/ci_tools/ci_interactions.py:6]: Yeah fair enough. There is no real logic to share here, but I'll look to do _something_ about that in a further PR. I kinda want to make `in_ci` return an enum value, but I haven't done the research to make that happen elegantly.
